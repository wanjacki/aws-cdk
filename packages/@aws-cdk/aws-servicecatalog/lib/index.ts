export * from './portfolio';
export * from './product';

// AWS::ServiceCatalog CloudFormation Resources:
export * from './servicecatalog.generated';
