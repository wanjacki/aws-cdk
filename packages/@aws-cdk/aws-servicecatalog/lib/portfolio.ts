/* eslint-disable */
import * as iam from '@aws-cdk/aws-iam';
import { IResource, Resource } from '@aws-cdk/core';
import { Construct } from 'constructs';
import { CfnPortfolio, CfnPortfolioPrincipalAssociation } from './servicecatalog.generated';

/**
 * A Service Catalog portfolio
 */
export interface IPortfolio extends IResource {

  /**
   * The ID of the portfolio.
   *
   * @attribute
   */
  readonly id: string;

  /**
   * The name of the portfolio 
   * 
   * @attribute
   */
  readonly portfolioName: string;

  /**
   * Associate portfolio with a principal
   * principal (Role/Group/User).
   */
  associatePrincipal(principal: iam.IRole): void;
}


/**
 * A reference to a portfolio
 */
export interface PortfolioAttributes {
  /**
  * The name of the portfolo.
  */
  readonly portfolioName: string;
}

/**
* Represents a Service Catalog portfolio.
*/
abstract class PortfolioBase extends Resource implements IPortfolio {

  /**
  * The id of the portfolio
  */
  public abstract readonly id: string;

  /**
  * The name of the portfolio
  */
  public abstract readonly portfolioName: string;


  /**
  * Associate a principal with the portfolio
  *
  */
  public abstract associatePrincipal(principal: iam.IRole): void;    


 
}

  /**
 * Properties for a Portfolio
 */
export interface PortfolioProps {
    /**
     * Enforces a particular physical stream name.
     * @default <generated>
     */
    readonly displayName: string;


  
  
    /**
     * The number of hours for the data records that are stored in shards to remain accessible.
     * @default no
     */
    readonly providerName: string;
  
    /**
     * The number of shards for the stream.
     * @default no
     */
    readonly acceptLanguage?: string;
  
    /**
     * The kind of server-side encryption to apply to this stream.
     *
     * If you choose KMS, you can specify a KMS key via `encryptionKey`. If
     * encryption key is not specified, a key will automatically be created.
     *
     * @default - no
     */
    readonly description?: string;
  
  }

  /**
 * A Service Catalog portfolio
 */
export class Portfolio extends PortfolioBase {
  
    public readonly portfolioName: string;
    public readonly id: string;
  
    private readonly portfolio: CfnPortfolio;

  
    constructor(scope: Construct, id: string, props: PortfolioProps) {
      super(scope, id, {
        physicalName: props.displayName,
      });

  
      this.portfolio = new CfnPortfolio(this, 'Resource', {
         displayName: props.displayName,
         providerName: props.providerName
      });
  
      this.portfolioName = this.getResourceNameAttribute(props.displayName);
      this.id = this.getResourceNameAttribute(this.portfolio.ref);

    }
  
   public associatePrincipal(principal: iam.IRole) {
    new CfnPortfolioPrincipalAssociation(this, 'PortfolioPrincipalAssociation', {
        portfolioId: this.portfolio.ref,
        principalArn: principal.roleArn,
        principalType: 'IAM'
    })
   }

  }

