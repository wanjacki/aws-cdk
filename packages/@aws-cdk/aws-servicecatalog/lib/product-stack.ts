import * as crypto from 'crypto';
import * as fs from 'fs';
import * as path from 'path';
import * as cdk from '@aws-cdk/core';
import { ProductStackSynthesizer } from './private/product-stack-synthesizer';

// keep this import separate from other imports to reduce chance for merge conflicts with v2-main
// eslint-disable-next-line no-duplicate-imports, import/order
import { Construct } from 'constructs';

export class ProductDetail {
  public bucketName?: string;
  public productStackId: string;
  public productName: string;
  public productVersionName: string;
  public productDescription: string;
  public validateTemplate: boolean;
  public overwriteExistingVersion: boolean;

  constructor(productStackId: string) {
    this.productStackId = productStackId;
    this.productName = '';
    this.productVersionName = '';
    this.productDescription = '';
    this.validateTemplate = true;
    this.overwriteExistingVersion = true;
  }

  public getBucketName(): string {
    return cdk.Lazy.uncachedString({ produce: () => this.bucketName });
  }

  public setBucketName(bucketName: string) {
    this.bucketName = bucketName;
  }

  public setProductName(productName: string) {
    this.productName = productName;
  }
  public setProductVersionName(productVersionName: string) {
    this.productVersionName = productVersionName;
  }
  public setProductDescription(productDescription: string) {
    this.productDescription = productDescription;
  }
  public setValidateTemplate(validateTemplate: boolean) {
    this.validateTemplate = validateTemplate;
  }
  public setOverwriteExistingVersion(setOverwriteExistingVersion: boolean) {
    this.overwriteExistingVersion = setOverwriteExistingVersion;
  }
}

/**
 * A Service Catalog product stack, which is similar in form to a Cloudformation nested stack.
 * You can add the resources to this stack that you want to define for your service catalog product.
 *
 * This stack will not be treated as an independent deployment
 * artifact (won't be listed in "cdk list" or deployable through "cdk deploy"),
 * but rather only synthesized as a template and uploaded as an asset to S3.
 *
 */
export class ProductStack extends cdk.Stack {
  public readonly templateFile: string;
  private readonly _productDetail: ProductDetail;
  private _templateUrl?: string;
  private _parentStack: cdk.Stack;

  constructor(scope: Construct, id: string) {
    super(scope, id, {
      synthesizer: new ProductStackSynthesizer(),
    });

    this._parentStack = findParentStack(scope);
    this._productDetail = new ProductDetail(id);

    // this is the file name of the synthesized template file within the cloud assembly
    const uniqueId = `${cdk.Names.uniqueId(this)}`;
    this.templateFile = `${uniqueId}.product.template.json`;
  }

  /**
   * Fetch the product details.
   *
   * @internal
   */
  public _getProductDetail(): ProductDetail | undefined {
    return this._productDetail;
  }

  /**
   * Fetch the template URL.
   *
   * @internal
   */
  public _getTemplateUrl(): string {
    return cdk.Lazy.uncachedString({ produce: () => this._templateUrl });
  }

  /**
   * Synthesize the product stack template, overrides the `super` class method.
   *
   * Defines an asset at the parent stack which represents the template of this
   * product stack.
   *
   * @internal
   */
  public _synthesizeTemplate(session: cdk.ISynthesisSession): void {
    const cfn = JSON.stringify(this._toCloudFormation(), undefined, 2);
    const templateHash = crypto.createHash('sha256').update(cfn).digest('hex');

    const templateAssetParent = this._parentStack.synthesizer.addFileAsset({
      packaging: cdk.FileAssetPackaging.FILE,
      sourceHash: templateHash,
      fileName: this.templateFile,
    });

    this._templateUrl = templateAssetParent.httpUrl;
    this._productDetail.setBucketName(templateAssetParent.bucketName);

    if (this._productDetail.overwriteExistingVersion != false) {
      this.writeToContext(templateHash, this._productDetail);
    }

    fs.writeFileSync(path.join(session.assembly.outdir, this.templateFile), cfn);
  }

  private writeToContext(templateHash: string, productDetail: ProductDetail) {
    const contextFileName = 'cdk.context.json';
    let contextJsonMap = {
      autoVersioningMap: {
        [productDetail.productName]: {
          [productDetail.productStackId]: {
            [productDetail.productVersionName]: {
              templateHash: templateHash,
              description: productDetail.productDescription,
              validateTemplate: productDetail.validateTemplate,
            },
          },
        },
      },
    };
    if (fs.existsSync(contextFileName)) {
      let contextJson = fs.readFileSync(contextFileName);
      contextJsonMap = JSON.parse(contextJson.toString());
      if (contextJsonMap.autoVersioningMap == undefined) {
        contextJsonMap.autoVersioningMap = {
          [productDetail.productName]: {
            [productDetail.productStackId]: {
              [productDetail.productVersionName]: {
                templateHash: templateHash,
                description: productDetail.productDescription,
                validateTemplate: productDetail.validateTemplate,
              },
            },
          },
        };
      } else if (contextJsonMap.autoVersioningMap[productDetail.productName] == undefined) {
        contextJsonMap.autoVersioningMap[productDetail.productName] = {
          [productDetail.productStackId]: {
            [productDetail.productVersionName]: {
              templateHash: templateHash,
              description: productDetail.productDescription,
              validateTemplate: productDetail.validateTemplate,
            },
          },
        };
      } else if (contextJsonMap.autoVersioningMap[productDetail.productName][productDetail.productStackId] == undefined) {
        contextJsonMap.autoVersioningMap[productDetail.productName][productDetail.productStackId] = {
          [productDetail.productVersionName]: {
            templateHash: templateHash,
            description: productDetail.productDescription,
            validateTemplate: productDetail.validateTemplate,
          },
        };
      } else {
        contextJsonMap.autoVersioningMap[productDetail.productName][productDetail.productStackId][productDetail.productVersionName] = {
          templateHash: templateHash,
          description: productDetail.productDescription,
          validateTemplate: productDetail.validateTemplate,
        };
      }
    }
    const contextJsonOutput = JSON.stringify(contextJsonMap);
    fs.writeFileSync('cdk.context.json', contextJsonOutput);
  }
}

/**
 * Validates the scope for a product stack, which must be defined within the scope of another `Stack`.
 */
function findParentStack(scope: Construct): cdk.Stack {
  try {
    const parentStack = cdk.Stack.of(scope);
    return parentStack as cdk.Stack;
  } catch (e) {
    throw new Error('Product stacks must be defined within scope of another non-product stack');
  }
}
