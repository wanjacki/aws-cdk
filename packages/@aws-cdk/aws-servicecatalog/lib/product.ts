/* eslint-disable */
import { IResource, Resource, Tag } from '@aws-cdk/core';
import { Construct } from 'constructs';
import { CfnCloudFormationProduct } from './servicecatalog.generated';


/** docstring*/
export interface IProduct extends IResource {

  /**
   * The id of the product
   *
   * @attribute
   */
  readonly id: string;


  /**
   * The name of the product
   *
   * @attribute
   */
  readonly name: string;

}


/**
* Represents a Service Catalog CloudFormation Product.
*/
abstract class ProductBase extends Resource implements IProduct {

  /**
  * The logical id of the product
  */
  public abstract readonly id: string;

  /**
  * The name of the product
  */
  public abstract readonly name: string;


 
}



/**
 * 
 * 
 * docstring
 */
export interface ProvisioningArtifactProperties {


        /**
     * The name of the product.
     * @default
     */
    readonly description?: string;

    /**
     * The name of the product.
     * @default True
     */
    readonly disableTemplateValidation?: boolean;

    /**
     * The name of the product.
     */
    readonly templateUrl: string;

    /**
     * The name of the product.
     * @default
     */
    readonly name?: string;


}




  /**
 * Properties for a Cloudformation Product
 */
export interface ProductProps {

    /**
     * The name of the product.
     */
    readonly name: string;

     /**
     * The language code.
     * @default 'en'
     */
    readonly acceptLanguage?: string;
  
    /**
     * The description of the product.
     * @default
     */
    readonly description?: string;
  
    /**
     * The distributor of the product.
     * @default
     */
    readonly distributor?: string;

  
    /**
     * The owner of the product.
     *
     */
    readonly owner: string;


    /**
     * The configuration of the provisioning artifact (also known as a version).
     * 
     */
    readonly provisioningArtifacts: ProvisioningArtifactProperties[];

    /**
     * ReplaceProvisioningArtifacts
     * @default false
     *
     */
     readonly replaceProvisioningParameters?: boolean;

    /**
     * The support information about the product
     * @default 
     */
     readonly supportDescription?: string;

    /**
     * The contact email for product support.
     * @default
     */
     readonly supportEmail?: string;

    /**
     * The contact URL for product support.
     * @default
     */
     readonly supportUrl?: string;

    /**
     * One or more tags.
     * @default 
     */
     readonly tags?: Tag[];
  
  }

  /**
 * A Service Catalog Cloudformation Product. 
 * @resource AWS::ServiceCatalog::CloudFormationProduct
 */
export class Product extends ProductBase {
  
    public readonly name: string;
    public readonly id: string;  
    private readonly product: CfnCloudFormationProduct;


    

  
    constructor(scope: Construct, id: string, props: ProductProps) {
        super(scope,id, {}
        );




  
      this.product = new CfnCloudFormationProduct(this, 'Resource', {
        acceptLanguage: props.acceptLanguage,
        description: props.description,
        distributor: props.distributor,
        name: props.name,
        owner: props.owner,
        provisioningArtifactParameters: this.generateProvisioningArtifactParameters(props),
        replaceProvisioningArtifacts: props.replaceProvisioningParameters,
        supportDescription: props.supportDescription,
        supportEmail: props.supportEmail,
        supportUrl: props.supportUrl,
        tags: props.tags
    
      });
  
      this.name = this.getResourceNameAttribute(props.name);
      this.id = this.getResourceNameAttribute(this.product.ref);



    }

    


    private generateProvisioningArtifactParameters(props: ProductProps): any[] {
    return props.provisioningArtifacts.map(pa => {
      return {
        name : pa.name,
        description : pa.description,
        disableTemplateValidation: pa.disableTemplateValidation,
        info : {
          LoadTemplateFromURL: pa.templateUrl
        }

      
    }});
  

 

  };

}