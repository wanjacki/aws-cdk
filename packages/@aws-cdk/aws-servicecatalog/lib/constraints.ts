/* eslint-disable */

import * as iam from '@aws-cdk/aws-iam';
import { IProduct } from './product';
import { AcceptLanguage } from './portfolio';
import { IRole } from '@aws-cdk/aws-iam';
import { CfnRule } from '@aws-cdk/core';

/**
 * Properties for LaunchRoleConstraint.
 */
export interface ConstraintOptions {
  /**
  * The language code.
  * @default 'en'
  */
  readonly acceptLanguage?: AcceptLanguage;

  /**
  * The description of the constraint.
  * @default ''
  */
  readonly description?: string;

  /**
  * A reference to product
  */
  readonly product: IProduct;
}


/**
 * Properties for LaunchNotificationConstraint.
 */
export interface LaunchNotificationConstraintProps extends ConstraintOptions {
  /**
  * A list of notification ARNs.
  */
  readonly notificationArns: string[];
}

/**
 * Properties for LaunchRoleConstraint.
 */
export interface LaunchRoleConstraintProps extends ConstraintOptions {
  /**
  * A reference to a Role
  */
  readonly role: iam.IRole;
}


/**
 * Properties for LaunchTemplateConstraint.
 */
export interface LaunchTemplateConstraintProps extends ConstraintOptions {
  /**
  * The rules for the template constraint
  * @default '{}'
  */
  readonly rules?: CfnRule;
}

/**
 * Properties for ResourceUpdateConstraint.
 */
export interface ResourceUpdateConstraintProps extends ConstraintOptions {
  /**
  * Toggle for if users should be allowed to change tags
  * @default true
  */
  readonly tagUpdateOnProvisionedProductAllowed?: boolean;
}

/**
 * Properties for StackSetConstraint.
 */
export interface StackSetConstraintProps extends ConstraintOptions {
  /**
  * One or more AWS accounts that will have access to the provisioned product
  */
  readonly accountList: string[];
  /**
  * One or more AWS Regions where the provisioned product will be available.
  */
  readonly regionList: string[];
  /**
  * ADmin Role
  */
  readonly adminRole: IRole;
  /**
  * Execution Role
  */
  readonly executionRole: IRole;
  /**
  * Toggle for Permission to create, update, and delete stack instances.
  * @default true
  */
  readonly stackInstanceControlAllowed?: boolean;
}

