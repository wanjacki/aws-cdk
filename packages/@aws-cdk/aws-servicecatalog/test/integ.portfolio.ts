import * as iam from '@aws-cdk/aws-iam';
import { App, Stack } from '@aws-cdk/core';
import { Portfolio } from '../lib';

const app = new App();
const stack = new Stack(app, 'integ-servicecatalog-portfolio');

const role = new iam.Role(stack, 'TestRole', {
  assumedBy: new iam.AccountRootPrincipal(),
});

const portfolio = new Portfolio(stack, 'TestPortfolio', {
  displayName: 'TestPortfolio',
  providerName: 'TestProvider',
});

portfolio.associatePrincipal(role);

app.synth();