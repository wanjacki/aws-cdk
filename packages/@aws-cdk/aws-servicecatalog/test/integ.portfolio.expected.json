{
    "Resources": {
      "TestRole6C9272DF": {
        "Type": "AWS::IAM::Role",
        "Properties": {
          "AssumeRolePolicyDocument": {
            "Statement": [
              {
                "Action": "sts:AssumeRole",
                "Effect": "Allow",
                "Principal": {
                  "AWS": {
                    "Fn::Join": [
                      "",
                      [
                        "arn:",
                        {
                          "Ref": "AWS::Partition"
                        },
                        ":iam::",
                        {
                          "Ref": "AWS::AccountId"
                        },
                        ":root"
                      ]
                    ]
                  }
                }
              }
            ],
            "Version": "2012-10-17"
          }
        }
      },
      "TestPortfolio4AC794EB": {
        "Type": "AWS::ServiceCatalog::Portfolio",
        "Properties": {
          "DisplayName": "TestPortfolio",
          "ProviderName": "TestProvider"
      }
    },
    "TestPortfolioPortfolioPrincipalAssociation7F3391C8": {
        "Type": "AWS::ServiceCatalog::PortfolioPrincipalAssociation",
        "Properties": {
          "PortfolioId": {"Ref": "TestPortfolio4AC794EB"},
          "PrincipalType": "IAM",
          "PrincipalARN":{
            "Fn::GetAtt": [
              "TestRole6C9272DF",
              "Arn"
            ]
          }
        }
    }
     
}
}